src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSyncBranchResolverActor.scala

```diff
@@ -0,0 +1,273 @@
+    with FastSyncBranchResolver
+    with PeerListSupportNg
+    with Timers {
+
+  import FastSyncBranchResolverActor._
+  import FastSyncBranchResolver._
+  import BinarySearchSupport._
+
+  private val recentHeadersSize: Int = syncConfig.blockHeadersPerRequest
+
+  private val recentBlocksSearch: RecentBlocksSearch = new RecentBlocksSearch(blockchain)
+
+  override def receive: Receive = waitingForPeerWithHighestBlock
+
+  private def waitingForPeerWithHighestBlock: Receive = handlePeerListMessages orElse { case StartBranchResolver =>
+    getPeerWithHighestBlock match {
+      case Some(PeerWithInfo(peer, _)) => requestRecentBlockHeaders(peer, blockchain.getBestBlockNumber())
+      case None =>
+        log.info("Waiting for peers, rescheduling StartBranchResolver")
+        timers.startSingleTimer(RestartTimerKey, StartBranchResolver, 1.second)
+    }
+  }
+
+  private def waitingForRecentBlockHeaders(
+      masterPeer: Peer,
+      bestBlockNumber: BigInt,
+      requestHandler: ActorRef
+  ): Receive =
+    handlePeerListMessages orElse {
+      case ResponseReceived(peer, BlockHeaders(headers), timeTaken) if peer == masterPeer =>
+        if (headers.size == recentHeadersSize) {
+          log.debug("Received {} block headers from peer {} in {} ms", headers.size, masterPeer.id, timeTaken)
+          handleRecentBlockHeadersResponse(headers, masterPeer, bestBlockNumber)
+        } else {
+          handleInvalidResponse(peer, requestHandler)
+        }
+      case RequestFailed(peer, reason) => handleRequestFailure(peer, sender(), reason)
+      case Terminated(ref) if ref == requestHandler => handlePeerTermination(masterPeer, ref)
+    }
+
+  private def waitingForBinarySearchBlock(
+      searchState: SearchState,
+      blockHeaderNumberToSearch: BigInt,
+      requestHandler: ActorRef
+  ): Receive = {
+    handlePeerListMessages orElse {
+      case ResponseReceived(peer, BlockHeaders(headers), durationMs) if peer == searchState.masterPeer =>
+        context.unwatch(requestHandler)
+        headers match {
+          case childHeader :: Nil if childHeader.number == blockHeaderNumberToSearch =>
+            log.debug(ReceivedBlockHeaderLog, blockHeaderNumberToSearch, peer.id, durationMs)
+            handleBinarySearchBlockHeaderResponse(searchState, childHeader)
+          case _ =>
+            log.warning(ReceivedWrongHeaders, blockHeaderNumberToSearch, headers.map(_.number))
+            handleInvalidResponse(peer, requestHandler)
+        }
+      case RequestFailed(peer, reason) => handleRequestFailure(peer, sender(), reason)
+      case Terminated(ref) if ref == requestHandler => handlePeerTermination(searchState.masterPeer, ref)
+      case Terminated(_) => () // ignore
+    }
+  }
+
+  private def requestRecentBlockHeaders(masterPeer: Peer, bestBlockNumber: BigInt): Unit = {
+    val requestHandler = sendGetBlockHeadersRequest(
+      masterPeer,
+      fromBlock = childOf(bestBlockNumber - recentHeadersSize),
```

Comment:
Minor: if bestBlockNumber is `0` this may request for blocks starting from-ve numbers.